Support ember cli 3.2

jacobq opened the pull request:
This is a cleaned-up version of https://github.com/felixrieseberg/ember-electron/pull/354
Please review; I'd love for ember-cli v3.2 support to land.

Closes #353

jacobq commented:
Uh oh, any idea why TravisCI isn't picking this up?

stopdropandrew commented on node-tests/acceptance/end-to-end-test.js:
```diff
@@ -97,7 +97,7 @@ describe('end-to-end', function() {
         process.chdir('ee-test-app');
 
-        return ember('install', `ember-electron@${path.join(packageTmpDir, 'package')}`);
+        return ember('install', `ember-electron@file:${path.join(packageTmpDir, 'ember-electron.tgz')}`);
```
Yes, because npm >5 symlinks packages instead via file which was causing module resolution to fail. https://medium.com/@alex_young/npm-5-and-file-urls-3c3631f7367c

bendemboski replied:
Ah, gotcha. Then let's re-tar the package directory into a new temp tarball and use that so it picks up the version-cache-busting.

jacobq replied:
@bendemboski I'm not sure how to do that. Is that something you or @stopdropandrew would be willing to do? (I'm done `git push -f`ing now -- fixed `.travis.yml`)

bendemboski replied:
I think something like this:

```javascript
return ember('new', 'ee-test-app').then(() => {
  process.chdir('ee-test-app');

  return run('tar', ['-czf', 'ember-electron-cachebust.tgz', '-C', packageTmpDir, 'package']);
}).then(() => {
  return ember('install', `ember-electron@file:${path.join(packageTmpDir, 'ember-electron-cachebust.tgz')}`);
});
```

jacobq replied:
So we're talking about code [in `node-tests/acceptance/end-to-end-test.js`](https://github.com/jacobq/ember-electron/blob/support-ember-cli-3.2/node-tests/acceptance/end-to-end-test.js#L56), right?
Sorry, I think I'm not following you guys. Why do we need to make a second tarball instead of just using the first? Are they different?

bendemboski replied:
Yes, that's correct. The reason I think we need to do it is because [here](https://github.com/jacobq/ember-electron/blob/support-ember-cli-3.2/node-tests/acceptance/end-to-end-test.js#L53-L62) we package up the app, then untar it so we can change the version in `package.json`. If we don't do that, then it would have the same version as the latest released version of `ember-electron.

At least when I wrote this, `yarn` would look at the version in `package.json`, and if it had a cached version of `ember-electron` with the same version, it would use the cached version and ignore the tarball.

The cached version could be the latest release installed into some other node project somewhere from the npm registry, or a previous version of the local package, like if you're doing local development and run the tests, then make a change, then run the same tests. So it was very problematic.

Perhaps this behavior doesn't currently happen with `npm`, and maybe not even with `yarn` any more, but it still seems sketchy to install a package from a tarball whose version exactly matches that package's version in the npm registry -- just asking for some caching behavior to screw us.

So that's why after packaging it up, we untar it and change the version.

jacobq replied:
OK, I think I get it now. We only use the extracted+modified contents of that tarball, but now we're using it directly, which may not be reliable due to potential cache issues.  I still have a couple questions though:
* Would you be opposed to this code being added in the same section where the package is packed, unpacked, and modified? (It makes sense to me to do it there even though one could argue that since it's only used in the `with npm` test right now we shouldn't do it until needed.)
* I know [`yarn pack` compresses the archive](https://yarnpkg.com/lang/en/docs/cli/pack/) and `.tgz`/`.tar.gz` is common, but since we're just making a temporary archive locally that we're going to immediately read back, I would prefer to just use `tar` ([`npm install` appears to support this](https://docs.npmjs.com/cli/install)) for performance / efficiency.

felixrieseberg commented on .travis.yml:
```diff
@@ -1,13 +1,20 @@
-  - "6.0"
-  - "7.0"
+addons:
+  chrome: stable
```
I had no idea you could do that 👏

jacobq replied:
[Yup](https://docs.travis-ci.com/user/chrome), but I only knew from riding the [coattails of ember-cli](https://github.com/ember-cli/ember-cli/pull/7161).

jacobq commented:
Uh oh, looks like I just went from all green to all red with just a couple lines of code (#skillz). The error message is confusing to me though. Is this because of my changes or because of a transient npm/yarn infrastructure problem? e.g.
* https://ci.appveyor.com/project/felixrieseberg/ember-electron/build/1.0.228/job/ud719hqk9dtuq8f0?fullLog=true#L368
* https://travis-ci.org/felixrieseberg/ember-electron/jobs/403080377#L1284

jacobq commented:
AppVeyor: only node 6.x passes, Travis: all except node 6.x pass. I can't make this stuff up :scream: :sob: :laughing:

jacobq commented:
Any idea what is causing [this `apt` issue](https://travis-ci.org/felixrieseberg/ember-electron/jobs/403172894#L582)? I suppose I can try escaping / quoting `g++-4.8` as [that may be a problem](https://stackoverflow.com/questions/30946960/the-command-sudo-apt-get-install-python-dev-python3-dev-failed-and-exited-wit#30950169), but why does it seem intermittent (e.g. only failed on node 10 job)?

jacobq commented:
More [crazy-making](https://travis-ci.org/felixrieseberg/ember-electron/jobs/403201676#L961):
```
Missing npm packages: 
Package: ember-electron
  * Specified: file:///tmp/tmp-5597c9HY8hfVEyIR/ember-electron-cachebust.tar
  * Installed: ../../tmp-5597c9HY8hfVEyIR/ember-electron-cachebust.tar
```

jacobq commented:
Where does this [`⠋ cleaning up⠋ cleaning up`](https://travis-ci.org/felixrieseberg/ember-electron/jobs/403218031#L736) junk come from? [`ember-cli`](https://github.com/ember-cli/ember-cli/blob/689bdbfee75ef92a33a86466669062bfb7507531/lib/models/builder.js#L183)? `yarn` (but we set `yarn config set no-progress`)?

bendemboski commented:
It's `ember-cli` -- when it finished building it prints `cleaning up...` -- you can see it by just running `ember build`

jacobq commented:
OK, [found](https://github.com/ember-cli/ember-cli/blob/master/lib/models/project.js#L763) [it](https://github.com/ember-cli/console-ui). Seems like it's making a progress spinner though, which tends to clutter the output in CI. I would've thought that's automatically disable when `CI` env var is present.

jacobq commented:
Same weirdness w/ `.tgz` :cry: Anyone know what's going on? This message is coming from `ember-cli` I think, not `npm` so maybe I need to check into that logic? (but "it worked before", which is the part that's killing me).
```
Missing npm packages: 
Package: ember-electron
  * Specified: file:../../tmp-5702nftN66mbim4B/ember-electron-cachebust.tgz
  * Installed: file:/tmp/tmp-5702nftN66mbim4B/ember-electron-cachebust.tgz
```

bendemboski commented:
Look like a relative/absolute path problem. Try using `path.join()` instead of `path.resolve()`?

jacobq commented:
@bendemboski I [started with `.join` and changed to `.resolve`](https://github.com/jacobq/ember-electron/commit/f0a38015cc66f86936cdafc1c5dae355191801cf) in hopes of fixing :joy:

jacobq commented:
I am tempted to just set [`SKIP_DEPENDENCY_CHECKER `](https://github.com/quaertym/ember-cli-dependency-checker/blob/master/lib/dependency-checker.js#L39) in the CI env vars 🤥

jacobq commented:
...40 CI builds later... I have a good feeling about this 😅

jacobq commented:
@bendemboski Would you mind taking another look? I've implemented the change you requested (cache-busting tar ball) and believe I've worked-out the snags encountered along the way. In short:

* [`ember-cli-dependency-checker`](https://github.com/quaertym/ember-cli-dependency-checker) has a [bug](https://github.com/quaertym/ember-cli-dependency-checker/issues/89) that causes it to mistakenly report that a package installed via tar ball is not satisfying the specified version (e.g. due to equivalent but not identically represented paths).
* I submitted a [patch PR](https://github.com/quaertym/ember-cli-dependency-checker/pull/90) to fix this and temporarily changed our test code to use that fork of `ember-cli-dependency-checker`.
* An unidentified problem with `npm` version 4 and earlier seems to prevent the installation from tar ball or verification of correctly installed dependencies from succeeding. Since `npm` 6 supports `node` 6 I've modified the CI configuration files to use `npm` 6 in the `node` 6 jobs (keeping the default versions for 8 [npm@5] and 10 [npm@6]).

stopdropandrew commented:
Looks good to me, thanks for the persistence this one had my head spinning.

jacobq commented:
@stopdropandrew You betcha (as we are accused of saying here in MN). Thanks for your help illuminating the bizarre `npm` sym-link/require behavior. Teamwork fo show 🥇

jacobq commented:
Whoops, looks like my last squash [doesn't get the latest `ember-cli-dependency-checker`](https://ci.appveyor.com/project/felixrieseberg/ember-electron/build/1.0.255/job/7lxw9gf3kcnwejhm#L1912).

jacobq commented:
Everything was passing then all I did was change from pinning `ember-cli-dependency-checker` to my fork commit back to using the release since the patch was merged, and now [the bug is back](https://ci.appveyor.com/project/felixrieseberg/ember-electron/build/1.0.262/job/94fiu237fkjbf093#L737). :sob:

jacobq commented on .appveyor.yml:
```diff
@@ -8,11 +8,14 @@ environment:
   - ps: Install-Product node $env:nodejs_version $env:platform
+  - cmd: IF "%nodejs_version%"=="6" npm install -g npm@6
+  - npm --version
   - npm install yarn -g
```
If we care, we can take this out now since [AppVeyor ships with yarn 1.7.0](https://www.appveyor.com/updates/2018/06/08/).

jacobq commented on node-tests/acceptance/end-to-end-test.js:
```diff
@@ -75,7 +75,7 @@ describe('end-to-end', function() {
       process.chdir(tmpDir);
 
-      return ember('new', 'ee-test-app', '--yarn').then(() => {
+      return ember('new', 'ee-test-app', '--yarn', '--skip-git', '--no-welcome').then(() => {
```
I forgot to mention this in the commit message, but I took the liberty of adding these parameters here since they shouldn't hurt anything (we're not using the git repository created for the new app nor the welcome page), and skipping these should improve test performance a little bit (less file I/O by skipping repo and fewer messages cluttering the log by not having the welcome page package installed).

jacobq commented on node-tests/acceptance/end-to-end-test.js:
```diff
@@ -94,10 +94,16 @@ describe('end-to-end', function() {
+        const packageJsonFileWithPath = path.join(process.cwd(), 'package.json');
+        const packageJson = readJsonSync(packageJsonFileWithPath);
+        //packageJson.devDependencies['ember-cli-dependency-checker'] = '^2.2.1';
+        packageJson.devDependencies['ember-cli-dependency-checker'] = 'quaertym/ember-cli-dependency-checker#^2.2.1';
```
If someone could explain it to me, I'd love to know why the commented-out line did not work but this one does. They should be equivalent since that is the main/canonical repository for this addon. It seems like an `npm` bug, but I don't want to investigate it further as it has already been too much of a time sink.

jacobq commented on .appveyor.yml:
```diff
@@ -1,15 +1,23 @@
 - x64
+
 environment:
-  nodejs_version: "6"
```
As much as I like being thorough and testing all LTS & current node releases, this does blow up the test time. Since we are doing all 3 on Travis (in parallel FTW) we could probably go back to just testing minimum supported node version on AppVeyor.

jacobq commented:
I just pushed a commit to try taking the `ember-cli-dependency-checker` hack out to see if it works like it should now. (Hoping it was a transient npm & caching thing, though I suspect it will fail.)

bendemboski commented:
Looks good to me overall. Two additional items that you can decide whether to address now or not:

1. `testem-electron.js` was originally based off of `testem.js` in the `ember-cli` app blueprint. It's been updated in some minor cosmetic ways, so we might consider updating `testem-electron.js` both at the root of the repo and in `blueprints/` to match.
2. `ember-cli` has moved to a single `.eslintrc` as the root of the app since `overrides` are now supported, but `ember-electron`'s blueprint is still creating `.eslintrc` files in subdirectories. It would take a little doing to make `ember-electron`'s blueprint modify the root `.eslintrc`, but shouldn't be too bad...but I'd also be happy getting this filed as a separate issue to address later since this PR has already been such an adventure!

jacobq commented:
@bendemboski I'll take another peek at that to see if anything looks out-of-date/sync. When I looked before I think I was just checking that [the breaking change regarding `args`](https://github.com/ember-cli/ember-cli/pull/7560) didn't affect it. Regarding 2, earlier today I created [an issue for ESlint improvement](https://github.com/felixrieseberg/ember-electron/issues/357), and I think I'd this would fit nicely into a separate PR that closes that.

jacobq commented:
Oh, cool, it sounds like the [`--headless` option may work for electron](https://github.com/electron/electron/issues/228#issuecomment-359385793). We might be able to drop the whole `Xvfb` work-around then.

jacobq commented on testem-electron.js:
```diff
@@ -1,14 +1,29 @@
+      ci: [
+        // --no-sandbox may be needed when running inside a container
+        process.env.CI ? '--no-sandbox' : null,
+        '--headless',
```
This doesn't seem to work. e.g. `ember electron --- --headless` still launches a visible window.

jacobq commented:
:rocket: ?

bendemboski commented:
🎉 nice work @jacobq! Perfect timing for me -- I'm just getting to upgrading my Electron app to the latest Ember, so thanks a bunch!

bendemboski commented:
Hey @felixrieseberg can you cut a new release so this code is out there for folks wanting to run against `ember-cli@3.2`?

felixrieseberg commented:
@bendemboski @jacobq Done, just published 2.9.0!
